keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
+  uid: abcd
+parameters:
+  - counter:
+      required: 100
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list
+and watch the custom claim parameter resources. This is necessary because it is
+not known in advance which CRDs will be used by DRA drivers. When deploying a
+driver, suitable role bindings for
+[`system:kube-scheduler`](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles)
+have to be added which grant read, watch and list permissions for the vendor
+CRD.
+
+The scheduler does not need to interpret the custom data in those CRDs. It
+still needs to read them for two purposes:
+- Ensure that the in-tree claim parameters match the vendor claim parameters
+  reference by the claim and are not out-dated.
+- Include a snapshot of claim and class parameters in the allocation result,
+  to ensure that configuration parameters are available when the DRA driver
+  needs them.
+
+The types that the scheduler needs to watch may change at runtime as classes
+get added or removed. To avoid repeated GET operations for those CRDs, dynamic
+informers will have to be used that get started and stopped as needed.
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+	...
+
+	// If (and only if) allocation of claims using this class is handled
+	// via numeric parameters, then ParameterTypes describes those
+	// parameters. A driver might support multiple different CRDs for
+	// its claims.
+	ParameterTypes []SemanticParameterType
+}
+
+// SemanticParameterType describes one resource that may be used for
+// claim parameters of a class which uses semantic parameters.
+type SemanticParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResources
+
+For each node, one or more NodeResources objects gets created:
+
+```
+// NodeResources provides information about available
+// resources on individual nodes.
+type NodeResources struct {
```

Comment:
Probably shouldn't be plural, based on API conventions, even though it contains multiple resources. I almost wonder if it should be called NodeResourceSlice, since it captures a subset of resources, though sharded by driver, it is a slice like EndpointSlice.

Reply:
There are (a few) other types which have the same singular and plural form. It's not common, but allowed.

I'm okay with using `NodeResourceSlice` instead. It nicely emphasizes that there are more of those per node.

Reply:
ResourceClaimParameters now is another example where IMHO the plural form makes sense for a single object because it contains multiple parameters.

I don't care strongly though.